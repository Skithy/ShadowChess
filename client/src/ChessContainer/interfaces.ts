import { ChessPiece, Team, Colour } from './constants'

export class Coord {
	readonly x: number
	readonly y: number

	static compareCoord = (c1: Coord, c2: Coord) => c1.x === c2.x && c1.y === c2.y

	constructor (x: number, y: number) {
		this.x = x
		this.y = y
	}

	compare = (c: Coord): boolean => Coord.compareCoord(c, this)
	getKey = (): number => this.y * 8 + this.x
	getCoords = (): number[] => [this.x, this.y]
}

export interface ITileState {
	chessPiece: ChessPiece
	team: Team
	coord: Coord
	fogOfWar?: boolean
}

export type BoardState = ITileState[][]

export interface IGameState {
	boardState: BoardState
	playerTurn: Team
	inCheck: boolean,
	ghostTile?: ITileState
}

export interface IChessIconSet {
	pawn: any // tslint:disable-line:no-any
	bishop: any // tslint:disable-line:no-any
	knight: any // tslint:disable-line:no-any
	rook: any // tslint:disable-line:no-any
	queen: any // tslint:disable-line:no-any
	king: any // tslint:disable-line:no-any
}

export interface ITeamInfo {
	colour: Colour
}
