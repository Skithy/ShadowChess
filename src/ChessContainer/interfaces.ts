import { ChessPiece, Team } from './constants'

export class Coord {
	readonly x: number
	readonly y: number

	static compareCoord = (c1: Coord, c2: Coord) => c1.x === c2.x && c1.y === c2.y

	constructor (x: number, y: number) {
		this.x = x
		this.y = y
	}

	compare = (c: Coord): boolean => Coord.compareCoord(c, this)
	getKey = (): number => this.y * 8 + this.x
	getCoords = (): number[] => [this.x, this.y]
}

export interface ITileState {
	chessPiece: ChessPiece
	team: Team
	coord: Coord
	fogOfWar?: boolean
}

export type BoardState = ITileState[][]

export interface IGameState {
	boardState: BoardState
	playerTurn: Team
}
