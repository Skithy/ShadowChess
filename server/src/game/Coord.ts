export default class Coord {
	public static compareCoord = (c1: Coord, c2: Coord): boolean => c1.x === c2.x && c1.y === c2.y

	public readonly x: number
	public readonly y: number

	constructor(x: number, y: number) {
		this.x = x
		this.y = y
	}

	public compare = (c: Coord): boolean => Coord.compareCoord(c, this)
	public getCoords = (): number[] => [this.x, this.y]
	public copy = (): Coord => new Coord(this.x, this.y)
}
