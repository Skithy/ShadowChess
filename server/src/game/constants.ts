export enum GameStatus {
	Game = 'Game',
	Paused = 'Paused',
	AwaitP1 = 'AwaitP1',
	AwaitP2 = 'AwaitP2',
}

export enum ChessPiece {
	Empty = '',
	Pawn = 'Pawn',
	Bishop = 'Bishop',
	Knight = 'Knight',
	Rook = 'Rook',
	Queen = 'Queen',
	King = 'King',
}

export enum Player {
	None = -1,
	One = 0,
	Two = 1,
}

export enum Colour {
	Aqua = 'Aqua',
	Black = 'Black',
	Blue = 'Blue',
	Green = 'Green',
	Grey = 'Grey',
	Orange = 'Orange',
	Purple = 'Purple',
	Red = 'Red',
	White = 'White',
	Yellow = 'Yellow',
}

export enum ColourCode {
	Aqua1 = '#16a085',
	Aqua2 = '#1abc9c',
	Black1 = '#2c3e50',
	Black2 = '#34495e',
	Blue1 = '#2980b9',
	Blue2 = '#3498db',
	Green1 = '#27ae60',
	Green2 = '#2ecc71',
	Grey1 = '#7f8c8d',
	Grey2 = '#95a5a6',
	Orange1 = '#d35400',
	Orange2 = '#e67e22',
	Purple1 = '#8e44ad',
	Purple2 = '#9b59b6',
	Red1 = '#c0392b',
	Red2 = '#e74c3c',
	Yellow1 = '#f39c12',
	Yellow2 = '#f1c40f',
	White1 = '#bdc3c7',
	White2 = '#ecf0f1',
}
