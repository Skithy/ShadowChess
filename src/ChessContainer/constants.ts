export enum ChessPiece {
	Empty,
	Pawn,
	Bishop,
	Knight,
	Rook,
	Queen,
	King
}

export enum Team {
	None,
	A,
	B
}