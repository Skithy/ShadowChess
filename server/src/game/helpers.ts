import Coord from './Coord'
import { Grid, ITileState } from './interfaces'

export const newGrid = (size: number): Grid => (
	Array.from(Array(size)).map((row, y): ITileState[] =>
		Array.from(Array(size)).map((tile, x): ITileState =>
			({ coord: new Coord(x, y) }),
		),
	)
)
