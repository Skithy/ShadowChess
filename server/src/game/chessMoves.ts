import { ChessPiece, Player } from './constants'
import Coord from './Coord'
import { Grid, IPieceInfo, ITileState } from './interfaces'

export const isValid = (x: number): boolean => x >= 0 && x < 8
const isFree = (tile: ITileState, fogOfWar?: boolean): boolean => !!(!tile.pieceInfo || (tile.fogOfWar && fogOfWar))

const addValidMove = (moves: Coord[], newTile: ITileState, player: Player, fogOfWar?: boolean): boolean => {
	// Valid if in fog of war - continues moving
	if (fogOfWar && newTile.fogOfWar) {
		moves.push(newTile.coord)
		return true
	}

	// Piece in the way - stops moving
	if (newTile.pieceInfo) {
		if (newTile.pieceInfo.player !== player) {
			moves.push(newTile.coord)
		}
		return false
	}

	// Nothing in the way - continues moving
	moves.push(newTile.coord)
	return true
}

const findPawnMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = piece.coord.getCoords()
	// Todo: Enpasse

	// Single-step and double-step moves
	const ssY = (piece.player === Player.One) ? y - 1 : y + 1
	const dsY = (piece.player === Player.One) ? y - 2 : y + 2

	// Check if single-step move is possible
	const ssTile = grid[ssY][x]
	if (isFree(ssTile, fogOfWar)) {
		moves.push(ssTile.coord)

		// Check if double-step move is possible
		if (!piece.hasMoved) {
			const dsTile = grid[dsY][x]
			if (isFree(dsTile, fogOfWar)) {
				moves.push(dsTile.coord)
			}
		}
	}

	// Diagonal capture - Able to move if forward-diagonal contains enemy piece
	for (const newX of [x - 1, x + 1]) {
		if (isValid(newX) && grid[ssY][newX].pieceInfo && grid[newX][ssY].pieceInfo!.player !== piece.player) {
			moves.push(new Coord(newX, ssY))
		}
	}

	return moves
}

const findBishopMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = piece.coord.getCoords()

	let tl = true
	let tr = true
	let bl = true
	let br = true
	for (let i = 1; i < 8; i++) {
		const xl = x - i
		const xr = x + i
		const yt = y - i
		const yb = y + i

		// Check bounds
		if (!isValid(xl)) {
			tl = false
			bl = false
		}
		if (!isValid(xr)) {
			tr = false
			br = false
		}
		if (!isValid(yt)) {
			tl = false
			tr = false
		}
		if (!isValid(yb)) {
			bl = false
			br = false
		}

		// Top Left
		if (tl) {
			const newTile = grid[yt][xl]
			tl = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Top Right
		if (tr) {
			const newTile = grid[yt][xr]
			tr = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Bottom Left
		if (bl) {
			const newTile = grid[yb][xl]
			bl = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Bottom Right
		if (br) {
			const newTile = grid[yb][xr]
			br = addValidMove(moves, newTile, piece.player, fogOfWar)
		}
	}

	return moves
}

const findKnightMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = piece.coord.getCoords()

	const xl1 = x - 1
	const xl2 = x - 2
	const xr1 = x + 1
	const xr2 = x + 2
	const yt1 = y - 1
	const yt2 = y - 2
	const yb1 = y + 1
	const yb2 = y + 2

	const coords = [
		[xl2, yt1], [xl1, yt2], [xr1, yt2], [xr2, yt1],
		[xl2, yb1], [xl1, yb2], [xr1, yb2], [xr2, yb1],
	]

	for (const [newX, newY] of coords) {
		if (isValid(newX) && isValid(newY)) {
			addValidMove(moves, grid[newY][newX], piece.player, fogOfWar)
		}
	}

	return moves
}

const findRookMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = piece.coord.getCoords()

	let top = true
	let bottom = true
	let left = true
	let right = true
	for (let i = 1; i < 8; i++) {
		// Top
		if (top && isValid(y - i)) {
			const newTile = grid[y - i][x]
			top = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Bottom
		if (bottom && isValid(y + i)) {
			const newTile = grid[y + i][x]
			bottom = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Left
		if (left && isValid(x - i)) {
			const newTile = grid[y][x - i]
			left = addValidMove(moves, newTile, piece.player, fogOfWar)
		}

		// Right
		if (right && isValid(x + i)) {
			const newTile = grid[y][x + i]
			right = addValidMove(moves, newTile, piece.player, fogOfWar)
		}
	}

	return moves
}

const findQueenMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = [
		...findBishopMoves(piece, grid, fogOfWar),
		...findRookMoves(piece, grid, fogOfWar),
	]
	return moves
}

const findKingMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = piece.coord.getCoords()

	// Todo: castle
	const xl = x - 1
	const xr = x + 1
	const yt = y - 1
	const yb = y + 1
	const coords = [
		[xl, yt], [x, yt], [xr, yt],
		[xl, y],           [xr, y],
		[xl, yb], [x, yb], [xr, yb],
	]

	for (const [newX, newY] of coords) {
		if (isValid(newX) && isValid(newY)) {
			addValidMove(moves, grid[newY][newX], piece.player, fogOfWar)
		}
	}

	return moves
}

export const findAvailableMoves = (piece: IPieceInfo, grid: Grid, fogOfWar?: boolean): Coord[] => {
	switch (piece.chessPiece) {
		case ChessPiece.Empty:
			return []
		case ChessPiece.Pawn:
			return findPawnMoves(piece, grid, fogOfWar)
		case ChessPiece.Bishop:
			return findBishopMoves(piece, grid, fogOfWar)
		case ChessPiece.Knight:
			return findKnightMoves(piece, grid, fogOfWar)
		case ChessPiece.Rook:
			return findRookMoves(piece, grid, fogOfWar)
		case ChessPiece.Queen:
			return findQueenMoves(piece, grid, fogOfWar)
		case ChessPiece.King:
			return findKingMoves(piece, grid, fogOfWar)
		default:
			return []
	}
}
