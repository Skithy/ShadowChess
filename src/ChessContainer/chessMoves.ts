import { ITileState, BoardState, Coord } from './interfaces'
import { ChessPiece, Team } from './constants'

export const isValid = (x: number) => x >= 0 && x < 8

const addValidMove = (moves: Coord[], tile: ITileState, newTile: ITileState, fogOfWar?: boolean): boolean => {
	if (fogOfWar && newTile.fogOfWar) {
		moves.push(newTile.coord)
		return true
	}

	if (newTile.chessPiece) {
		if (newTile.team !== tile.team) {
			moves.push(newTile.coord)
		}
		return false
	}

	moves.push(newTile.coord)
	return true
}

const findPawnMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = tile.coord.getCoords()
	// Todo: Enpasse
	if (tile.team === Team.B) {
		// Single-step
		const ssTile = boardState[y + 1][x]
		if (!ssTile.chessPiece || (fogOfWar && ssTile.fogOfWar)) {
			moves.push(ssTile.coord)
			// Double-step
			const dsTile = boardState[y + 2][x]
			if (y === 1 && (!dsTile.chessPiece || (fogOfWar && dsTile.fogOfWar))) {
				moves.push(dsTile.coord)
			}
		}

		// Diagonal capture
		if (isValid(x - 1) && boardState[y + 1][x - 1].team === Team.A) {
			moves.push(new Coord(x - 1, y + 1))
		}
		if (isValid(x + 1) && boardState[y + 1][x + 1].team === Team.A) {
			moves.push(new Coord(x + 1, y + 1))
		}
	} else {
		// Single-step
		const ssTile = boardState[y - 1][x]
		if (!ssTile.chessPiece || (fogOfWar && ssTile.fogOfWar)) {
			moves.push(ssTile.coord)
			// Double-step
			const dsTile = boardState[y - 2][x]
			if (y === 6 && (!dsTile.chessPiece || (fogOfWar && dsTile.fogOfWar))) {
				moves.push(dsTile.coord)
			}
		}

		// Diagonal capture
		if (isValid(x - 1) && boardState[y - 1][x - 1].team === Team.B) {
			moves.push(new Coord(x - 1, y - 1))
		}
		if (isValid(x + 1) && boardState[y - 1][x + 1].team === Team.B) {
			moves.push(new Coord(x + 1, y - 1))
		}
	}

	return moves
}

const findBishopMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = tile.coord.getCoords()

	let tl = true
	let tr = true
	let bl = true
	let br = true
	for (let i = 1; i < 8; i++) {
		// Check bounds
		if (!isValid(x - i)) {
			tl = false
			bl = false
		}
		if (!isValid(x + i)) {
			tr = false
			br = false
		}
		if (!isValid(y - i)) {
			tl = false
			tr = false
		}
		if (!isValid(y + i)) {
			bl = false
			br = false
		}

		// Top Left
		if (tl) {
			const newTile = boardState[y - i][x - i]
			tl = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Top Right
		if (tr) {
			const newTile = boardState[y - i][x + i]
			tr = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Bottom Left
		if (bl) {
			const newTile = boardState[y + i][x - i]
			bl = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Bottom Right
		if (br) {
			const newTile = boardState[y + i][x + i]
			br = addValidMove(moves, tile, newTile, fogOfWar)
		}
	}

	return moves
}

const findKnightMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = tile.coord.getCoords()

	const xl1 = x - 1
	const xl2 = x - 2
	const xr1 = x + 1
	const xr2 = x + 2
	const yt1 = y - 1
	const yt2 = y - 2
	const yb1 = y + 1
	const yb2 = y + 2
	
	const coords = [
		[xl2, yt1], [xl1, yt2], [xr1, yt2], [xr2, yt1],
		[xl2, yb1], [xl1, yb2], [xr1, yb2], [xr2, yb1],
	]

	for (let [newX, newY] of coords) {
		if (isValid(newX) && isValid(newY)) {
			addValidMove(moves, tile, boardState[newY][newX], fogOfWar)
		}
	}

	return moves
}

const findRookMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = tile.coord.getCoords()

	let top = true
	let bottom = true
	let left = true
	let right = true
	for (let i = 1; i < 8; i++) {
		// Top
		if (top && isValid(y - i)) {
			const newTile = boardState[y - i][x]
			top = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Bottom
		if (bottom && isValid(y + i)) {
			const newTile = boardState[y + i][x]
			bottom = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Left
		if (left && isValid(x - i)) {
			const newTile = boardState[y][x - i]
			left = addValidMove(moves, tile, newTile, fogOfWar)
		}

		// Right
		if (right && isValid(x + i)) {
			const newTile = boardState[y][x + i]
			right = addValidMove(moves, tile, newTile, fogOfWar)
		}
	}

	return moves
}

const findQueenMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = [
		...findBishopMoves(tile, boardState, fogOfWar),
		...findRookMoves(tile, boardState, fogOfWar),
	]
	return moves
}

const findKingMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	const moves: Coord[] = []
	const [x, y] = tile.coord.getCoords()

	// Todo: castle
	const xl = x - 1
	const xr = x + 1
	const yt = y - 1
	const yb = y + 1
	const coords = [
		[xl, yt], [x, yt], [xr, yt],
		[xl, y], [], [xr, y],
		[xl, yb], [x, yb], [xr, yb],
	]

	for (let [newX, newY] of coords) {
		if (isValid(newX) && isValid(newY)) {
			addValidMove(moves, tile, boardState[newY][newX], fogOfWar)
		}
	}

	return moves
}

export const findAvailableMoves = (tile: ITileState, boardState: BoardState, fogOfWar?: boolean): Coord[] => {
	switch (tile.chessPiece) {
		case ChessPiece.Empty:
			return []
		case ChessPiece.Pawn:
			return findPawnMoves(tile, boardState, fogOfWar)
		case ChessPiece.Bishop:
			return findBishopMoves(tile, boardState, fogOfWar)
		case ChessPiece.Knight:
			return findKnightMoves(tile, boardState, fogOfWar)
		case ChessPiece.Rook:
			return findRookMoves(tile, boardState, fogOfWar)
		case ChessPiece.Queen:
			return findQueenMoves(tile, boardState, fogOfWar)
		case ChessPiece.King:
			return findKingMoves(tile, boardState, fogOfWar)
		default:
			return []
	}
}
