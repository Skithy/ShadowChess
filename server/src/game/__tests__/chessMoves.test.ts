import { findAvailableMoves } from '../chessMoves'
import { ChessPiece, Player } from '../constants'
import Coord from '../Coord'
import { newGrid } from '../helpers'
import { Grid, IPieceInfo } from '../interfaces'

const newPiece = (chessPiece: ChessPiece, player: Player, coord: Coord): IPieceInfo => ({
	chessPiece,
	coord,
	hasMoved: false,
	isAlive: true,
	player,
	possibleMoves: [],
})

describe('chessMoves', () => {
	let grid: Grid
	beforeEach(() => {
		grid = newGrid(8)
	})

	describe('pawn', () => {
		describe('can go forward 2 moves if first move', () => {
			test('Player 1', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 6))
				const moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(2)
				expect(moves).toContainEqual([2, 5])
				expect(moves).toContainEqual([2, 4])
				expect(moves).toMatchSnapshot()
			})

			test('Player 2', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 1))
				const moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(2)
				expect(moves).toContainEqual([2, 2])
				expect(moves).toContainEqual([2, 3])
				expect(moves).toMatchSnapshot()
			})
		})

		describe('can go forward 1 move if not first move', () => {
			test('Player 1', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 5))
				pawn.hasMoved = true
				const moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(1)
				expect(moves).toContainEqual([2, 4])
				expect(moves).toMatchSnapshot()
			})

			test('Player 2', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 2))
				pawn.hasMoved = true
				const moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(1)
				expect(moves).toContainEqual([2, 3])
				expect(moves).toMatchSnapshot()
			})
		})

		describe('cant go forward if blocked by any piece', () => {
			test('Player 1 - single move', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 6))
				grid[5][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 5))

				expect(findAvailableMoves(pawn, grid).length).toBe(0)
				grid[5][2].pieceInfo!.player = Player.Two
				expect(findAvailableMoves(pawn, grid).length).toBe(0)
			})

			test('Player 1 - double move', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 6))
				grid[4][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 4))

				expect(findAvailableMoves(pawn, grid).length).toBe(1)
				grid[4][2].pieceInfo!.player = Player.Two
				expect(findAvailableMoves(pawn, grid).length).toBe(1)
			})

			test('Player 2 - single move', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 1))
				grid[2][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 2))

				let moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(0)

				grid[2][2].pieceInfo!.player = Player.One
				moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(0)
			})

			test('Player 2 - double move', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 1))
				grid[3][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 3))

				let moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(1)

				grid[3][2].pieceInfo!.player = Player.One
				moves = findAvailableMoves(pawn, grid).map((coord) => coord.getCoords())
				expect(moves.length).toBe(1)
			})
		})

		describe('can go forward if blocked but in fog of war', () => {
			test('Player 1 - single space', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 6))
				grid[5][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 5))
				grid[5][2].fogOfWar = true
				expect(findAvailableMoves(pawn, grid, true).length).toBe(2)
			})

			test('Player 1 - double space', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 6))
				grid[4][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 4))
				grid[4][2].fogOfWar = true
				expect(findAvailableMoves(pawn, grid, true).length).toBe(2)
			})

			test('Player 2 - single space', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 1))
				grid[2][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 2))
				grid[2][2].fogOfWar = true
				expect(findAvailableMoves(pawn, grid, true).length).toBe(2)
			})

			test('Player 2 - double space', () => {
				const pawn = newPiece(ChessPiece.Pawn, Player.Two, new Coord(2, 1))
				grid[3][2].pieceInfo = newPiece(ChessPiece.Pawn, Player.One, new Coord(2, 3))
				grid[3][2].fogOfWar = true
				expect(findAvailableMoves(pawn, grid, true).length).toBe(2)
			})
		})

		// describe('can take diagonal pieces', () => {

		// })

	})
})
