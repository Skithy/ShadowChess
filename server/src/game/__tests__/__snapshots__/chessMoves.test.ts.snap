// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chessMoves pawn can go forward 1 move if not first move Player 1 1`] = `
Array [
  Array [
    2,
    4,
  ],
]
`;

exports[`chessMoves pawn can go forward 1 move if not first move Player 2 1`] = `
Array [
  Array [
    2,
    3,
  ],
]
`;

exports[`chessMoves pawn can go forward 2 moves if first move Player 1 1`] = `
Array [
  Array [
    2,
    5,
  ],
  Array [
    2,
    4,
  ],
]
`;

exports[`chessMoves pawn can go forward 2 moves if first move Player 2 1`] = `
Array [
  Array [
    2,
    2,
  ],
  Array [
    2,
    3,
  ],
]
`;
